hoomd/md/TwoStepRATTLEBDGPU.cu

```diff
@@ -0,0 +1,312 @@
+
+    This kernel must be launched with enough dynamic shared memory per block to read in d_gamma
+*/
+extern "C" __global__
+void gpu_rattle_brownian_step_one_kernel(Scalar4 *d_pos,
+                                  int3 *d_image,
+                                  const BoxDim box,
+                                  const Scalar *d_diameter,
+                                  const unsigned int *d_rtag,
+                                  const unsigned int *d_groupTags,
+                                  const unsigned int nwork,
+                                  const Scalar4 *d_net_force,
+                                  const Scalar3 *d_f_brownian,
+                                  const Scalar3 *d_gamma_r,
+                                  Scalar4 *d_orientation,
+                                  Scalar4 *d_torque,
+                                  const Scalar3 *d_inertia,
+                                  Scalar4 *d_angmom,
+                                  const Scalar *d_gamma,
+                                  const size_t n_types,
+                                  const bool use_alpha,
+                                  const Scalar alpha,
+                                  const unsigned int timestep,
+                                  const unsigned int seed,
+                                  const Scalar T,
+                                  const bool aniso,
+                                  const Scalar deltaT,
+                                  unsigned int D,
+                                  const bool d_noiseless_r,
+                                  const unsigned int offset)
+    {
+    HIP_DYNAMIC_SHARED( char, s_data)
+
+    Scalar3 *s_gammas_r = (Scalar3 *)s_data;
+    Scalar *s_gammas = (Scalar *)(s_gammas_r + n_types);
+
+    if (!use_alpha)
+        {
+        // read in the gamma (1 dimensional array), stored in s_gammas[0: n_type] (Pythonic convention)
+        for (int cur_offset = 0; cur_offset < n_types; cur_offset += blockDim.x)
+            {
+            if (cur_offset + threadIdx.x < n_types)
+                s_gammas[cur_offset + threadIdx.x] = d_gamma[cur_offset + threadIdx.x];
+            }
+        __syncthreads();
+        }
+
+    // read in the gamma_r, stored in s_gammas_r[0: n_type], which is s_gamma_r[0:n_type]
+
+    for (int cur_offset = 0; cur_offset < n_types; cur_offset += blockDim.x)
+        {
+        if (cur_offset + threadIdx.x < n_types)
+            s_gammas_r[cur_offset + threadIdx.x] = d_gamma_r[cur_offset + threadIdx.x];
+        }
+    __syncthreads();
+
+    // determine which particle this thread works on (MEM TRANSFER: 4 bytes)
+    int local_idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (local_idx < nwork)
+        {
+        const unsigned int group_idx = local_idx + offset;
+
+        // determine the particle to work on
+        unsigned int tag = d_groupTags[group_idx];
```

Comment:
Don't use `d_groupTags` here. Use the group member index array that ParticleGroup provides coupled with the tag information from ParticleData. See TwoStepBD which gets the index and tag of the particle with:
```
unsigned int idx = d_group_members[group_idx];
unsigned int ptag = d_tag[idx];
```